File: client/lib/controllers/vehicle_controller.dart
import 'package:client/models/vehicle.dart';
import 'package:client/urls.dart';
import 'package:dio/dio.dart';

import '../adapters/dio_client.dart';

class VehicleController{
    VehicleController._();

    static void addVehicle(Vehicle vehicleDetails, {void Function()? onSuccess, void Function()? onFailed}){
        onSuccessful(Response<dynamic> response){
            onSuccess?.call();
        }

        onError(Response<dynamic> response){
          onFailed?.call();
        }

       

        DioClient(onError: onError, onSuccess: onSuccessful, url: addcar, authenticatedRequest: true).post(vehicleDetails.toJson());
    }

    static void getVehicles({void Function(List<Vehicle> vehicles)? onSuccess, void Function()? onFailed}){
        onSuccessful(Response<dynamic> response){
          
            onSuccess?.call((response.data as List).map((vehicle) => Vehicle.fromJson(vehicle)).toList());
        }

        onError(Response<dynamic> response){

        }

       

        DioClient(onError: onError, onSuccess: onSuccessful, url: addcar, authenticatedRequest: true).get();
    }
}

File: web/src/controllers/vehicle-controller.ts
import AxiosClient from 'adapters/ApiClient';
import { AddVehicleFailedFunction, AddVehicleSuccessFunction, GetVehiclesFailedFunction, GetVehiclesSuccessFunction } from 'types/vehicle';
import { addvehicle } from 'urls';
import { VehicleDetails } from './../interfaces/vehicle';
//@ts-ignore
import {cloneDeep} from 'lodash';

export const addVehicle = (details: VehicleDetails, onSuccess: AddVehicleSuccessFunction, onFailed: AddVehicleFailedFunction) => {
    const onSuccessful = (response:any) => {

        const {brand, color, make, model, plateNumber, _id:id} = response.data;
        onSuccess({
            brand, make, color, model, plateNumber, id
        }) 
    }

    const onError = (error:any) => {

    }

    const onTimeout = () => {

    }

    new AxiosClient(addvehicle, onSuccessful, onError, onTimeout, true).postWithoutConversion(details);
    
}

export const getVehicles = (onSuccess: GetVehiclesSuccessFunction, onFailed: GetVehiclesFailedFunction) => {
    const onSuccessful = (response:any) => {
        onSuccess(cloneDeep((response.data)?.map?.((vehicle:any) => {
            const {brand, color, make, model, plateNumber, _id:id} = vehicle

            return {
                    brand, make, color, model, plateNumber, id
                }
        })))
        // const {brand, color, make, model, plateNumber, _id:id} = response.data;
        // onSuccess({
        //     brand, make, color, model, plateNumber, id
        // }) 
    }

    const onError = (error:any) => {

    }

    const onTimeout = () => {

    }

    new AxiosClient(addvehicle, onSuccessful, onError, onTimeout, true).get();
    
}